### PR TITLE
feat: improve clickhouse connection management

### LINKED ISSUE #18079: Clickhouse exporter not working with https
### Component(s)

_No response_

## Description

Hi, I cannot connect to a Clickhouse instance over HTTPS. It would be great if you could provide an example if I'm doing something wrong.

## Steps to Reproduce

Make a free account on https://clickhouse.com/ and try to connect the exporter with it.

### Collector version

0.70.0

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

```yaml
receivers:
  otlp:
    protocols:
      grpc:
      http:

processors:
  batch:
    send_batch_size: 4000
    send_batch_max_size: 4000
    timeout: 10s
  memory_limiter:
    # 80% of maximum memory up to 2G
    limit_mib: 1500
    # 25% of limit up to 2G
    spike_limit_mib: 512
    check_interval: 5s

exporters:
  clickhouse:
    endpoint: "https://$HOST:8443"
    database: otel
    logs_table_name: otel_logs
    traces_table_name: otel_traces
    username: foo
    password: bar
    connection_params:
      read_timeout: 20s
      write_timeout: 30s
      secure: true
      debug: true
    ttl_days: 3
    timeout: 20s
    sending_queue:
      queue_size: 100
    retry_on_failure:
      enabled: true
      initial_interval: 5s
      max_interval: 30s
      max_elapsed_time: 300s

extensions:
  health_check:

service:
  extensions: [health_check]
  pipelines:
    traces:
      receivers: [otlp]
      processors: [memory_limiter, batch]
      exporters: [clickhouse]
```

### Log output

```shell
2023-01-27T21:56:30Z   [info]Error: cannot build pipelines: failed to create "clickhouse" exporter, in pipeline "traces": cannot configure clickhouse traces exporter: create database:Post "https://YOUR_HOST:8443?database=default&default_format=Native&write_timeout=30s": dial tcp: lookup YOUR_HOST: i/o timeout
2023-01-27T21:56:30Z   [info]2023/01/27 21:56:30 collector server run finished with error: cannot build pipelines: failed to create "clickhouse" exporter, in pipeline "traces": cannot configure clickhouse traces exporter: create database:Post "https://YOUR_HOST:8443?database=default&default_format=Native&write_timeout=30s": dial tcp: lookup YOUR_HOST: i/o timeout
2023-01-27T21:56:31Z   [info]Starting clean up.
```

### Additional context

I tried it with `github.com/ClickHouse/clickhouse-go/v2` through `sql.Open` and it worked fine.

```go
	url := fmt.Sprintf(
		"https://%s:%d?username=%s&password=%s&secure", host, https_port, username, password)
```

### PR DESCRIPTION
**Pull request by StarpTech**
**Description:** <Describe what has changed.>

Improve clickhouse DSN parsing to support all possible configuration options and fixes TLS support.

This will also fix #18079

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

# TODO

- [x] Fixing tests

### CODE REVIEW

File: exporter/clickhouseexporter/exporter_logs.go
Comment (hanjm) [inline exporter/clickhouseexporter/exporter_logs.go:200]: it is nessary to switch to default database to create new database.
if the database not default and not exist, `cfg.buildDB` will return no such database error.

so last PR #18067 split `dsn` to `endpoint` `database` and other connection_params, make it easy to switch default database, replace is fragile (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16476).

File: exporter/clickhouseexporter/exporter_logs_test.go
Comment (hanjm) [inline exporter/clickhouseexporter/exporter_logs_test.go:96]: opentelemetry exporter prefer use the word `endpoint` instead of `addr` or `host`.
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13442#discussion_r980507122

File: exporter/clickhouseexporter/config.go
Comment (hanjm) [inline exporter/clickhouseexporter/config.go:110]: It is better to make connection pool as config settings.

Comment (StarpTech) [reply exporter/clickhouseexporter/config.go:110]: Done

File: exporter/clickhouseexporter/exporter_logs_test.go
Comment (StarpTech) [reply exporter/clickhouseexporter/exporter_logs_test.go:96]: Done

File: exporter/clickhouseexporter/exporter_logs.go
Comment (StarpTech) [reply exporter/clickhouseexporter/exporter_logs.go:200]: Hi, I see your point, but in that case, we don't manipulate strings. We overwrite the database field after parsing all options. The database name from the DSN URLs is ignored after all.

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [inline exporter/clickhouseexporter/README.md:264]: it should keep consistant, use`endpoint` instead of `addr` or `host`.

File: exporter/clickhouseexporter/exporter_logs.go
Comment (hanjm) [reply exporter/clickhouseexporter/exporter_logs.go:200]: LGTM, @dmitryax @StarpTech  please take a look.

File: exporter/clickhouseexporter/config.go
Comment (hanjm) [reply exporter/clickhouseexporter/config.go]: I think make it as config field, not just remove it. 

set `max_idle_conn` is useful when we want to use short connection for load balance.

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [inline exporter/clickhouseexporter/README.md:264]: let's keep `endpoint` for consistant.

Comment (StarpTech) [reply exporter/clickhouseexporter/README.md:264]: ok

**Comment by StarpTech** [comment]
Could someone enable CI for this?

File: exporter/clickhouseexporter/internal/exponential_histogram_metrics.go
Comment (Frapschen) [inline exporter/clickhouseexporter/internal/exponential_histogram_metrics.go:130]: `e.count` was added by:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b2f0015292f57228b8ffbfd2f8fe405ceb6e8de1/exporter/clickhouseexporter/internal/exponential_histogram_metrics.go#L191
so, I think `e.count*expHistogramValueCounts` does calculate the total count. please double check

File: exporter/clickhouseexporter/internal/gauge_metrics.go
Comment (Frapschen) [inline exporter/clickhouseexporter/internal/gauge_metrics.go:112]: please double check

File: exporter/clickhouseexporter/internal/gauge_metrics.go
Comment (Frapschen) [inline exporter/clickhouseexporter/internal/gauge_metrics.go:132]: should start with 11

File: exporter/clickhouseexporter/internal/histogram_metrics.go
Comment (Frapschen) [inline exporter/clickhouseexporter/internal/histogram_metrics.go:122]: please double check

File: exporter/clickhouseexporter/internal/summary_metrics.go
Comment (Frapschen) [inline exporter/clickhouseexporter/internal/summary_metrics.go:108]: please double check

**Comment by Frapschen** [comment]
I think approver want you to fix those bug separately.

File: exporter/clickhouseexporter/internal/gauge_metrics.go
Comment (StarpTech) [reply exporter/clickhouseexporter/internal/gauge_metrics.go:132]: Good catch. Fixed.

File: exporter/clickhouseexporter/internal/exponential_histogram_metrics.go
Comment (StarpTech) [reply exporter/clickhouseexporter/internal/exponential_histogram_metrics.go:130]: You're right; I didn't see that the value was already accumulated.

**Comment by StarpTech** [comment]
> I think approver want you to fix those bug separately.

See description. As soon as I get a response that my fixes can be merged reasonably, I will create separate PRs. I need the whole package, and it's currently much easier to build a custom otel binary.

**Review by Frapschen — COMMENTED** [review]
LGTM

**Comment by Frapschen** [comment]
> > I think approver want you to fix those bug separately.
> 
> See description. As soon as I get a response that my fixes can be merged reasonably, I will create separate PRs. I need the whole package, and it's currently much easier to build a custom otel binary.

Can you separate the bugfix PRs then I  can ping the approver on Slack to get those merged quickly?

**Comment by dmitryax** [comment]
@StarpTech can you please create separate PRs for all the issues with incorporated feedback from this PR? It'll be easier to review and faster to merge

**Comment by Frapschen** [comment]
@StarpTech I help your to separate them: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18288 https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18287 @dmitryax

**Comment by StarpTech** [comment]
@Frapschen thank you.

**Comment by StarpTech** [comment]
The last one  https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18289 @dmitryax PRs for all issues was created. As soon they are merged, I'll clean up this PR.

**Comment by Frapschen** [comment]
> @Frapschen thank you. Please add me as a co-author on merges.

sorry for one of PRs have got merged before I see this message.

**Comment by StarpTech** [comment]
Please remove the stale label. I need more time to merge all the changes.

**Comment by StarpTech** [comment]
Hi @dmitryax @hanjm I merged main and resolved all conflicts. Please take another look.

**Comment by StarpTech** [comment]
I'd appreciate some help. The CI is a beast https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4338134175/jobs/7574616510

**Comment by StarpTech** [comment]
Everything is green :smile:

File: exporter/clickhouseexporter/README.md
Comment (hanjm) [inline exporter/clickhouseexporter/README.md:320]: what is the reason to remove this section?

File: exporter/clickhouseexporter/config.go
Comment (hanjm) [inline exporter/clickhouseexporter/config.go:100]: It seems no need to make a copy.

File: exporter/clickhouseexporter/README.md
Comment (StarpTech) [reply exporter/clickhouseexporter/README.md:320]: It just bloats up the readme and requires managing it in two places. The scheme definition is in the code.

File: exporter/clickhouseexporter/config.go
Comment (StarpTech) [reply exporter/clickhouseexporter/config.go]: Done

**Comment by atoulme** [comment]
It’s a lot of changes, sorry. If it’s at all possible, would you please peel off a few of them into separate PRs? One for the clickhouse upgrade, one for the README change, to start.

File: exporter/clickhouseexporter/config.go
Comment (atoulme) [inline exporter/clickhouseexporter/config.go:63]: I don’t quite understand that change. Why change the error?

Comment (StarpTech) [reply exporter/clickhouseexporter/config.go:63]: I can revert it. The implementation did change a couple of times.

**Comment by StarpTech** [comment]
> It’s a lot of changes, sorry. If it’s at all possible, would you please peel off a few of them into separate PRs? One for the clickhouse upgrade, one for the README change, to start.

Hi @atoulme I think this is micromanagement. The README changes are made due to my modifications and the clickhouse upgrade is a patch. If the changes are not welcome, I can revert it. It's important for us that the DSN handling gets merged. The lack of response in the PR reviews prevents me from creating more PRs than necessary :disappointed:

**Comment by atoulme** [comment]
Guilty as charged! I'm actually a manager by trade. I would love to help but the changes are a bit on the large side. I am sure others can step in and help. Smaller focused PRs usually get a fast response (not just from me, from others too!). You can bring your PR at a SIG meeting for discussion.

File: exporter/clickhouseexporter/config.go
Comment (dmitryax) [inline exporter/clickhouseexporter/config.go:44]: Why do we need to move the params from this option to the endpoint if we parse them after that anyway? It's also a breaking change which is ok for the alpha component but still better to avoid if there are no strong reasons for that

File: cmd/configschema/go.mod
Comment (dmitryax) [inline cmd/configschema/go.mod:240]: nit: Let's avoid dependency upgrades if no functionality depends on that. It'll be handled by automated depebdabot updates anyway.

File: exporter/clickhouseexporter/README.md
Comment (dmitryax) [reply exporter/clickhouseexporter/README.md:320]: As a rule of thumb, it's always best to keep the PRs to one purpose and make unrelated changes in separate PRs. Otherwise, it's confusing to the reviewers who would ask "Why does improving clickhouse connection management need removing these blocks?"

File: exporter/clickhouseexporter/config.go
Comment (StarpTech) [reply exporter/clickhouseexporter/config.go:44]: The most common approach is to set query parameters on the [DSN](https://github.com/ClickHouse/clickhouse-go#dsn). We could have both, but it would have made it a bit more complicated. I'd like to use the endpoint-only approach.

File: cmd/configschema/go.mod
Comment (StarpTech) [reply cmd/configschema/go.mod:240]: When are deps upgraded? It would be odd to revert it because there is another way.

File: exporter/clickhouseexporter/README.md
Comment (StarpTech) [reply exporter/clickhouseexporter/README.md:266]: Yes, introduced by clickhouse driver.

Comment (atoulme) [reply exporter/clickhouseexporter/README.md:266]: Should `clickhouse` be in the README? also it looks like the default is `clickhouse://127.0.0.1:9000`?

File: exporter/clickhouseexporter/config.go
Comment (dmitryax) [reply exporter/clickhouseexporter/config.go:44]: I agree it should be one explicit way to configure the DSN. But it's inconsistent now. We ask for Database, Username, and Password separately, but not ConnectionParams.

Comment (StarpTech) [reply exporter/clickhouseexporter/config.go:44]: It's an improvement. I removed `ConnectionParams`. I kept `Username` and `Password` separate because it's handy when using a secret store. Some people store the full DSN and some only the credentials.

Comment (StarpTech) [reply exporter/clickhouseexporter/config.go:44]: We could also just go with the endpoint if you don't agree.

File: exporter/clickhouseexporter/README.md
Comment (StarpTech) [reply exporter/clickhouseexporter/README.md:266]: Yes, I will add it.

Comment (StarpTech) [reply exporter/clickhouseexporter/README.md:266]: Added

File: exporter/clickhouseexporter/config.go
Comment (StarpTech) [reply exporter/clickhouseexporter/config.go]: Restored.

Comment (StarpTech) [reply exporter/clickhouseexporter/config.go]: I restored the `ConnectionParams` option. The PR only contains the TLS fix with some internal improvements in how DSN is handled. Tests were added. Readme changes and dep upgrades are minor things. If you still think this is not mergeable, please help; otherwise, I will close the PR.

**Review by dmitryax — APPROVED** [review]

File: .chloggen/refactor-clickhouse-conn-management.yaml
Comment (dmitryax) [inline .chloggen/refactor-clickhouse-conn-management.yaml:2]: It's not a breaking change anymore, right?

File: exporter/clickhouseexporter/README.md
Comment (dmitryax) [inline exporter/clickhouseexporter/README.md:267]: Is the protocol part supposed to be repeated or typo?
```suggestion
  - tcp protocol `tcp://addr1:port,tcp://addr2:port` or TLS `tcp://addr1:port,addr2:port?secure=true`
  - http protocol `http://addr1:port,addr2:port` or https `https://addr1:port,addr2:port`
  - clickhouse protocol `clickhouse://addr1:port,addr2:port` or TLS `clickhouse://addr1:port,addr2:port?secure=true`
```

Comment (hanjm) [reply exporter/clickhouseexporter/README.md:267]: @StarpTech there are no clickhouse protocol, just http/https or native tcp, `clickhouse://` is same just same to `tcp:// `.

https://github.com/ClickHouse/clickhouse-go/blob/08b27884b899f587eb5c509769cd2bdf74a9e2a1/clickhouse_options.go#L303-L315

Comment (StarpTech) [reply exporter/clickhouseexporter/README.md:267]: But they sell it and it works https://github.com/ClickHouse/clickhouse-go#dsn What's your proposal?

File: .chloggen/refactor-clickhouse-conn-management.yaml
Comment (StarpTech) [reply .chloggen/refactor-clickhouse-conn-management.yaml:2]: correct

**Comment by dmitryax** [comment]
@StarpTech please rebase

**Comment by StarpTech** [comment]
@dmitryax done

**Review by atoulme — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

File: exporter/clickhouseexporter/README.md
Comment (et) [reply exporter/clickhouseexporter/README.md:320]: > The scheme definition is in the code.

Where can I find the schema definition in code? I was referencing the definition in the README (for some internal documentation) and I would like to point to the correct location.

Comment (cwegener) [reply exporter/clickhouseexporter/README.md:320]: > > The scheme definition is in the code.
> 
> Where can I find the schema definition in code? I was referencing the definition in the README (for some internal documentation) and I would like to point to the correct location.

For each table there is a `CREATE TABLE` statement like this:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/4f27accf111db991b87a19f658f375d076390669/exporter/clickhouseexporter/internal/gauge_metrics.go#L31-L64

Comment (et) [reply exporter/clickhouseexporter/README.md:320]: Thanks!

**Comment by polRk** [comment]
@StarpTech Why did you remove the Schema section?

**Comment by StarpTech** [comment]
@polRk Because we don't want to keep the schema synced in two places.

**Comment by polRk** [comment]
> @polRk Because we don't want to keep the schema synced in two places.

Where is the second place?

**Comment by StarpTech** [comment]
In code. Every schema is in its corresponding go file.

**Pull request merged**
